### DISCUSSION TITLE
Fit canvas to container

### DISCUSSION THREAD

**Discussion by tv42** [cetz-package/cetz#695] [Q&A]
If I've understood correctly, cetz canvas "interior coordinate" bounds are always the bounds of whatever is drawn there, and setting a unit value for the canvas `length` argument sets the physical on-page size of whatever `1` means inside the canvas; actual size comes indirectly from multiplying this unit size by `length`.

To get a fixed size out of cetz, it seems I need to say `length: 100%/canvas_interior_width`. And I can't seem to be able to figure out the canvas interior width programmatically -- my leftmost and rightmost content are text labels, and I don't directly control their size!

Trying to pull of some measure-then-rescale trick with `#layout(size => let m = measure(my-canvas))` leads to `cannot expand into infinite width` error from cetz canvas.

This makes it incredibly hard to consistently set diagrams to be e.g. 80% of the column width.

Is there any way to make this better?

**Comment by johannes-wolf** [comment]
Is this answered in https://github.com/cetz-package/cetz/discussions/689?

Using Typst' `measure` and `scale`:
```
#let c = cetz.canvas({
  import cetz.draw: *

  circle((1,1), fill: red)
  content((5,0), [Some test text is here. \ What do you think?])
  rotate(280deg)
  rect((1,3), (rel: (2, 2)), fill: blue)
})

#place(context {
  let (width, height) = measure(c)
  scale(10cm / width * 100%, place(c))
})
```

**Comment by tv42** [comment]
I think you just hardcoded the size of the container in that `10cm`? That needs to use an invocation of `layout` I think. Doable by itself, but it's misbehaving in other ways:

Using `place` makes the following content overlap the drawing. They seem to be doing something to the layout though, removing both instances of `place` makes the drawing shift to the left, inside the left margin.

Adding `float: true` to both `place` makes the following content not overlap the drawing, but now I've lost control of where in the text the drawing goes -- float moves it to bottom of page.

**Comment by tv42** [comment]
Ah, it's the `scale` that makes the drawing poke into the left margin. `scale(origin: left, ` makes it work correctly.

Here's what I now have:

```typst
#let diagram(canvas) = {
  context {
    let canvas-size = measure(canvas)
    [canvas width=#canvas-size.width]
    layout(size => {
      [parent width=#size.width]
      scale(origin: left, size.width/canvas-size.width * 100%, canvas)
    })
  }
}

#diagram(cetz.canvas(
  {
    import cetz.draw: *

    content((-5,0), [left])
    content((5,0), [right])
    }
  ))
```

The remaining annoyances are that just simple `scale` changes font sizes and stroke widths. It'd be much better to inform the `canvas` of the area it has available, and let stroke widths and text sizes like 8pt stay as 8pt.

**Comment by johannes-wolf** [comment]
You can also try the first option (without `scale`), I've mentioned in the linked issue. The problem is, `content` might not resize the bounds correctly.
